lib/semantics/type.cc

```diff
@@ -38,22 +39,253 @@ void DerivedTypeSpec::ReplaceScope(const Scope &scope) {
+  }
+
+  // Parameters of the most deeply nested "base class" come first when the
+  // derived type is an extension.
+  auto parameterNames{OrderParameterNames(typeSymbol_)};
+  auto parameterDecls{OrderParameterDeclarations(typeSymbol_)};
+  auto nextNameIter{parameterNames.begin()};
+  RawParameters raw{std::move(rawParameters_)};
+  for (auto &[maybeKeyword, value] : raw) {
+    SourceName name;
+    common::TypeParamAttr attr{common::TypeParamAttr::Kind};
+    if (maybeKeyword) {
+      name = maybeKeyword->v.source;
+      auto it{std::find_if(parameterDecls.begin(), parameterDecls.end(),
+          [&](const Symbol &symbol) { return symbol.name() == name; })};
+      if (it == parameterDecls.end()) {
+        messages.Say(name,
+            "'%s' is not the name of a parameter for derived type '%s'"_err_en_US,
+            name, typeSymbol_.name());
+      } else {
+        // Resolve the keyword's symbol
+        maybeKeyword->v.symbol = const_cast<Symbol *>(&it->get());
+        attr = it->get().get<TypeParamDetails>().attr();
+      }
+    } else if (nextNameIter != parameterNames.end()) {
+      name = *nextNameIter++;
+      auto it{std::find_if(parameterDecls.begin(), parameterDecls.end(),
+          [&](const Symbol &symbol) { return symbol.name() == name; })};
+      CHECK(it != parameterDecls.end());
+      attr = it->get().get<TypeParamDetails>().attr();
+    } else {
+      messages.Say(name_,
+          "Too many type parameters given for derived type '%s'"_err_en_US,
+          typeSymbol_.name());
+      break;
+    }
+    if (FindParameter(name)) {
+      messages.Say(name_,
+          "Multiple values given for type parameter '%s'"_err_en_US, name);
+    } else {
+      value.set_attr(attr);
+      AddParamValue(name, std::move(value));
+    }
+  }
+}
+
+void DerivedTypeSpec::EvaluateParameters(
+    evaluate::FoldingContext &foldingContext) {
+  if (evaluated_) {
+    return;
+  }
+  evaluated_ = true;
+  auto &messages{foldingContext.messages()};
+
+  // Fold the explicit type parameter value expressions first.  Do not
+  // fold them within the scope of the derived type being instantiated;
+  // these expressions cannot use its type parameters.  Convert the values
+  // of the expressions to the declared types of the type parameters.
+  auto parameterDecls{OrderParameterDeclarations(typeSymbol_)};
+  for (const Symbol &symbol : parameterDecls) {
+    const SourceName &name{symbol.name()};
+    if (ParamValue * paramValue{FindParameter(name)}) {
+      if (const MaybeIntExpr & expr{paramValue->GetExplicit()}) {
+        if (auto converted{evaluate::ConvertToType(symbol, SomeExpr{*expr})}) {
+          SomeExpr folded{
+              evaluate::Fold(foldingContext, std::move(*converted))};
+          if (auto *intExpr{std::get_if<SomeIntExpr>(&folded.u)}) {
+            paramValue->SetExplicit(std::move(*intExpr));
+            continue;
+          }
+        }
+        std::stringstream fortran;
+        fortran << *expr;
```

Comment:
I think this should use `expr->AsFortran(fortran);`. Even though they may be equivalent now, `AsFortran` implies a guarantee (i.e. that its output is Fortran) that `operator<<` doesn't.